File: src/main/java/com/mateusz/spring_boot_todos/models/Todo.java
package com.mateusz.spring_boot_todos.models;

import java.time.LocalDate;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonIgnore;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.validation.constraints.Future;
import jakarta.validation.constraints.Size;

@Entity
public class Todo {

    @Id
    @GeneratedValue
    private Integer id;

    @ManyToOne
    @JoinColumn(name = "username", referencedColumnName = "username")
    @JsonIgnore
    private User user;

    @JsonGetter("username")
    public String getUsername() {
        return this.user.getUsername();
    }

    @Size(min = 5)
    private String description;

    @Future
    private LocalDate targetDate;// LocalDate recommended over Date

    private boolean done;

    public Todo(Integer id, User user, String description, LocalDate targetDate, boolean done) {
        this.id = id;
        this.user = user;
        this.description = description;
        this.targetDate = targetDate;
        this.done = done;
    }

    public Todo() {
    }

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public LocalDate getTargetDate() {
        return targetDate;
    }

    public void setTargetDate(LocalDate targetDate) {
        this.targetDate = targetDate;
    }

    public boolean isDone() {
        return done;
    }

    public void setDone(boolean done) {
        this.done = done;
    }

    @Override
    public String toString() {
        return "Todo [id=" + id + ", user=" + user + ", description=" + description + ", targetDate=" + targetDate
                + ", done=" + done + "]";
    }

}

File: src/main/kotlin/com/mateusz/todos/kotlin_sboot_todos_backend/models/Todo.kt
package com.mateusz.todos.kotlin_sboot_todos_backend.models

import com.fasterxml.jackson.annotation.JsonGetter
import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.*
import jakarta.validation.constraints.Future
import jakarta.validation.constraints.Size
import java.time.LocalDate

@Entity
class Todo {
    @Id
    @GeneratedValue
    var id: Int? = null

    @ManyToOne
    @JoinColumn(name = "username", referencedColumnName = "username")
    @JsonIgnore
    var user: User? = null

    @get:JsonGetter("username")
    val username: String?
        get() = user!!.username

    var description: @Size(min = 5) String? = null

    var targetDate: @Future LocalDate? = null // LocalDate recommended over Date

    var done: Boolean = false

    constructor(id: Int?, user: User?, description: String?, targetDate: LocalDate?, done: Boolean) {
        this.id = id
        this.user = user
        this.description = description
        this.targetDate = targetDate
        this.done = done
    }

    constructor()

    override fun toString(): String {
        return ("Todo [id=" + id + ", user=" + user + ", description=" + description + ", targetDate=" + targetDate
                + ", done=" + done + "]")
    }
}